xtext/org.icyphy.linguafranca/src/lib/Python/pythontarget.c

```diff
@@ -0,0 +1,1141 @@
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+ * @section DESCRIPTION
+ * Implementation of functions defined in @see pythontarget.h
+ */
+
+#include "pythontarget.h"
+
+//////////// set Function(s) /////////////
+/**
+ * Set the value and is_present field of self which is of type
+ * LinguaFranca.port_capsule
+ * 
+ * Each LinguaFranca.port_capsule includes a void* pointer of 
+ * the C port (a.k.a. generic_port_instance_struct*).
+ * @see generic_port_capsule_struct in pythontarget.h
+ * 
+ * This function calls the underlying _LF_SET API.
+ * @see xtext/org.icyphy.linguafranca/src/lib/core/reactor.h
+ * 
+ * This function can be used to set any type of PyObject ranging from
+ * primitive types to complex lists and tuples. Moreover, this function
+ * is callable from Python target code by using port_name.out(value)
+ * 
+ * Some examples include
+ *  port_name.out("Hello")
+ *  port_name.out(5)
+ *  port_name.out(["Hello", 5 , (2.8, "X")])
+ * 
+ * The port type given in the Lingua Franca is only used as a "suggestion"
+ * as per Python's duck typing principles. The end-user is responsible for
+ * appropriately handling types on the recieveing end of this port.
+ * @param self The output port (by name) or input of a contained
+ *                 reactor in form input_name.port_name.
+ * @param args contains:
+ *      @param val The value to insert into the port struct.
+ */
+static PyObject* py_SET(PyObject *self, PyObject *args)
+{
+    generic_port_capsule_struct* p = (generic_port_capsule_struct*)self;
+    PyObject* val, *tmp;
+
+    if (!PyArg_ParseTuple(args, "O", &val))
+    {
+        PyErr_SetString(PyExc_TypeError, "Could not set objects.");
+        return NULL;
+    }
+
+    generic_port_instance_struct* port = PyCapsule_GetPointer(p->port, "port");
+    if(port == NULL)
+    {
+        fprintf(stderr, "Null pointer recieved.\n");
+        exit(1);
+    }
+    
+    if(val)
+    {   
+        tmp = port->value;
+        Py_INCREF(val);
+        _LF_SET(port, val);
+
+        // Also set the values for the port capsule.       
+        p->value = val;
+        p->is_present = true;
+        //Py_XDECREF(tmp); // Since value is allocated in Python, the Python garbage collector will manage and free this memory
```

Comment:
`Py_INCREF` in this case is necessary because we are introducing a new reference to the object. This way the Python garbage collector will not free the object during the reaction code execution. The `Py_XDECREF` is optional since the object is already garbage collected. We can force it to free sooner but then we have to perform an extra check for `Py_None`.